File: Retos/Reto #24 - CIFRADO CÉSAR [Fácil]/python/KevinED11.py
from string import ascii_uppercase
from abc import ABC, abstractmethod


class InvalidCharacterError(Exception):
    def __init__(self, message: str) -> None:
        super().__init__(message)

class InvalIdCipherError(Exception):
    pass

class InvalidDecrypterError(Exception):
    pass


class ICipher(ABC):
    @property
    @abstractmethod
    def cipher(self) -> str:
        pass
    

class IAlphabet(ABC):
    @abstractmethod
    def get(self) -> str:
        pass


class IDecrypter(ABC):
    @property
    @abstractmethod
    def decrypt(self) -> str:
        pass


class IConversor(ABC):
    @property
    @abstractmethod
    def get(self) -> dict[str, str]:
        pass


class EnglishAlphabet(IAlphabet):
    @staticmethod
    def get() -> str:
        return ascii_uppercase + " "


class SpanishAlphabet(IAlphabet):
    @staticmethod
    def get() -> str:
        return "abcdefghijklmnñopqrstuvwxyz ".upper()


class CesarCipherConversor(IConversor):
    def __init__(self, alphabet: IAlphabet, shift: int) -> None:
        self.__alphabet = alphabet.get()
        self.__shift = shift
    
    @property
    def get(self) -> dict[str, str]:
        letters = self.__alphabet
        cesar = {}
        
        for i, letter in enumerate(letters):
            cesar_position = (i + self.__shift) % len(letters)
            cesar_letter = letters[cesar_position]
            cesar[letter] = cesar_letter
        return cesar



# M

def convert_text(text: str, conversor: dict[str, str]) -> str:
    conversion = ""
    for letter in text:
           if letter not in conversor:
                raise InvalidCharacterError(f"Invalid character: {letter}")
           conversion += conversor[letter]

    return conversion


class CesarCipher(ICipher):
    def __init__(self, text: str, conversor: IConversor) -> None:
        self.__text = text.upper()
        self.__conversor = conversor.get
    
    @property
    def cipher(self) -> str:
        converted_text = convert_text(text=self.__text, conversor=self.__conversor)
        
        return converted_text


class CesarCipherDecrypter(IDecrypter):
    def __init__(self, cesar_text: str, conversor: IConversor) -> None:
        self.__cesar_text = cesar_text.upper()
        self.__conversor = conversor.get
    
    @property
    def decrypt(self) -> str:
        conversor = {v: k for k, v in self.__conversor.items()}
        converted_text = convert_text(text=self.__cesar_text, conversor=conversor)
        
        return converted_text


class Main:
    def __init__(self, cipher: ICipher, decrypter: IDecrypter) -> None:
        self.__cipher = cipher
        self.__decrypter = decrypter
        self.__post_init__()
    
    def __post_init__(self) -> None:
        if not issubclass(type(self.__cipher), ICipher):
            raise InvalidDecrypterError("Invalid cipher")
        if not issubclass(type(self.__decrypter), IDecrypter):
            raise InvalidDecrypterError("Invalid decrypter")

        
    def cipher(self) -> str:
        return self.__cipher.cipher
    
    def decrypt(self) -> str:
        return self.__decrypter.decrypt


if __name__ == "__main__":
    alphabet = EnglishAlphabet()
    conversor = CesarCipherConversor(alphabet, 3)
    try:
        cipher = CesarCipher("kevin duenas perro", conversor)
        decrypter = CesarCipherDecrypter(cipher.cipher, conversor)
        
        program = Main(cipher, decrypter)
        print(program.cipher())
        print(program.decrypt())
    except (InvalidCharacterError,
            InvalIdCipherError,
            InvalidDecrypterError) as err:
        print(err)

File: Retos/Reto #24 - CIFRADO CÉSAR [Fácil]/go/KevinED11.go
package main

import (
	"errors"
	"fmt"
	_ "fmt"
	"strings"
)

type IAlphabet interface {
	Get() string
}

type IChiper interface {
	Encrypt() (string, error)
}

type IConversor interface {
	GetMapConversion() *map[string]string
}

type IDecrypter interface {
	Decrypt() (string, error)
}

type EnglishAlphabet struct {
}

func (a *EnglishAlphabet) Get() string {
	return "abcdefghijklmnopqrstuvwxyz"
}

type CaesarCipherConversor struct {
	IAlphabet
	Shift int
}

func (c *CaesarCipherConversor) GetMapConversion() *map[string]string {
	conversor := make(map[string]string)
	alphabet := c.IAlphabet.Get()

	for i, letter := range alphabet {
		cesarPosition := (i + c.Shift) % len(alphabet)
		cesarLetter := string(alphabet[cesarPosition])
		conversor[string(letter)] = cesarLetter
	}

	return &conversor

}

type CaesarCipher struct {
	text string
	IConversor
}

func (c *CaesarCipher) Encrypt() (string, error) {
	var cesarText string
	conversor := *c.IConversor.GetMapConversion()

	text := strings.ToLower(c.text)

	for _, letter := range text {
		let, exists := conversor[string(letter)]
		if !exists {
			return "", errors.New("Invalid letter")
		}

		cesarText += let
	}

	return cesarText, nil

}

type CaesarCipherDecrypter struct {
	text string
	IConversor
}

func (c *CaesarCipherDecrypter) Decrypt() (string, error) {
	text := strings.ToLower(c.text)
	originalMap := *c.IConversor.GetMapConversion()
	newMap := make(map[string]string)
	var decryptedText string

	for k, v := range originalMap {
		newMap[v] = k
	}

	for _, letter := range text {
		let, exist := newMap[string(letter)]
		if !exist {
			return "", errors.New("Invalid letter")
		}

		decryptedText += let
	}

	return decryptedText, nil

}

type Main struct {
	IChiper
	IDecrypter
}

func (m *Main) Encrypt() string {
	result, err := m.IChiper.Encrypt()
	if err != nil {
		fmt.Println(err)
	}
	return result
}

func (m *Main) Decrypt() string {
	result, err := m.IDecrypter.Decrypt()
	if err != nil {
		fmt.Println(err)
	}
	return result
}

func main() {
	alphabet := EnglishAlphabet{}
	cipherConversor := CaesarCipherConversor{IAlphabet: &alphabet, Shift: 3}

	cipher := CaesarCipher{text: "Hello", IConversor: &cipherConversor}
	decrypter := CaesarCipherDecrypter{text: "khoor", IConversor: &cipherConversor}

	program := Main{IChiper: &cipher, IDecrypter: &decrypter}
	encryptedText := program.Encrypt()
	decryptedText := program.Decrypt()

	fmt.Println(encryptedText, decryptedText)

}
